### DISCUSSION TITLE
Support for Queue Selectors

### DISCUSSION THREAD

**Discussion by jlplasce** [rabbitmq/rabbitmq-jms-client#251] [General]
As of now, RMQSession still throw an exception when a new consumers to queues is created with selectors:

https://github.com/rabbitmq/rabbitmq-jms-client/blob/d8a716f7616cd960d16bdf6842a930ab24731b2a/src/main/java/com/rabbitmq/jms/client/RMQSession.java#L834

This is a definite use-case for JMS users trying to use this library to interface with RMQ

**Comment by michaelklishin** [comment]
If something hasn't been added in years there is often a reason for it: in the original protocol supported by RabbitMQ, "queue selectors" happen at a different stage that JMS assumes.

It will take a while to get to a truly protocol agnostic core and then some more time for this client to be migrated to, say, use a different protocol (potentially [streams](https://rabbitmq.com/streams.html)).

**Reply by acogoluegnes** [reply]
Message selection is implemented at the exchange level for [JMS topics](https://rabbitmq.github.io/rabbitmq-jms-client/3.x/stable/htmlsingle/index.html#jms-topic-support). A topic subscriber has its own AMQP queue and its optional message selector is known at creation time, so messages can be filtered out during the routing phase, not once they reached the destination, when they are about to be dispatched.

This would be a different story with queues, as the message next in line would need to be matched against the optional consumer message selector, and dispatched accordingly (to the consumer or to the next one).

This is the first time the support for message selectors in JMS queues has come up in more than 6 years of maintaining the RabbitMQ JMS client in my case.
